File: rust/integration-gateway/src/router.rs
use std::collections::HashMap;

use axum::extract::State;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::routing::post;
use axum::{Json, Router};
use common_metrics::inc;
use serde::{Deserialize, Serialize};
use serde_json::json;
use uuid::Uuid;

use crate::app_context::AppState;
use crate::audit::{self, AuditEvent};
use crate::auth::AuthedCaller;
use crate::integrations::DecryptedIntegration;
use crate::metrics_consts;

#[derive(Deserialize)]
pub struct FetchRequest {
    pub integration_ids: Vec<i64>,
}

#[derive(Serialize)]
pub struct FetchResponse {
    /// integration id (as a string key) -> decrypted integration, or null when the id doesn't
    /// exist or belongs to another team (indistinguishable on purpose).
    pub integrations: HashMap<String, Option<DecryptedIntegration>>,
}

/// Merge the credential API (which carries `AppState`) onto a stateless base router (health/index).
pub fn merge_api_routes(app: Router, state: AppState) -> Router {
    let api = Router::new()
        .route("/api/v1/credentials/fetch", post(fetch_credentials))
        .with_state(state);
    app.merge(api)
}

async fn fetch_credentials(
    State(state): State<AppState>,
    caller: AuthedCaller,
    Json(req): Json<FetchRequest>,
) -> Response {
    if req.integration_ids.len() > state.max_batch_size {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": format!("too many integration_ids (max {})", state.max_batch_size) })),
        )
            .into_response();
    }

    let request_id = Uuid::new_v4().to_string();

    let outcome = match state
        .service
        .get_for_team(caller.team_id, &req.integration_ids)
        .await
    {
        Ok(outcome) => outcome,
        Err(e) => {
            tracing::error!(error = %e, request_id, "integration fetch failed");
            inc(
                metrics_consts::FETCH_TOTAL,
                &[
                    ("caller".to_string(), caller.caller.clone()),
                    ("result".to_string(), "error".to_string()),
                ],
                1,
            );
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "internal error" })),
            )
                .into_response();
        }
    };

    // Preserve request order/keys: every requested id appears in the response, resolved or null.
    let mut integrations: HashMap<String, Option<DecryptedIntegration>> =
        HashMap::with_capacity(req.integration_ids.len());
    let mut resolved_ids: Vec<i64> = Vec::new();
    for id in &req.integration_ids {
        match outcome.resolved.get(id) {
            Some(v) => {
                resolved_ids.push(*id);
                integrations.insert(id.to_string(), Some((**v).clone()));
            }
            None => {
                integrations.insert(id.to_string(), None);
            }
        }
    }

    audit::emit(&AuditEvent {
        caller: &caller.caller,
        team_id: caller.team_id,
        requested: &req.integration_ids,
        resolved: &resolved_ids,
        cache_hits: outcome.cache_hits,
        db_loaded: outcome.db_loaded,
        request_id: &request_id,
    });

    inc(
        metrics_consts::FETCH_TOTAL,
        &[
            ("caller".to_string(), caller.caller.clone()),
            ("result".to_string(), "ok".to_string()),
        ],
        1,
    );

    (StatusCode::OK, Json(FetchResponse { integrations })).into_response()
}

File: nodejs/src/integration-gateway/router.ts
import { randomUUID } from 'crypto'
import express from 'ultimate-express'

import { logger } from '~/common/utils/logger'

import { emitAudit } from './audit'
import { GatewayAuth } from './auth'
import { IntegrationService } from './integration.service'
import { recordFetch } from './metrics'
import { DecryptedIntegration } from './types'

export interface GatewayRouterDeps {
    service: IntegrationService
    auth: GatewayAuth
    maxBatchSize: number
}

interface FetchResponseBody {
    // integration id (string key) -> decrypted integration, or null when the id doesn't exist or
    // belongs to another team (indistinguishable on purpose).
    integrations: Record<string, DecryptedIntegration | null>
}

/**
 * Build the credential API router. `POST /api/v1/credentials/fetch` authenticates with a scoped
 * JWT (never the internal API secret) and returns every requested id, resolved or null.
 */
export function createGatewayRouter(deps: GatewayRouterDeps): express.Router {
    const router = express.Router()

    router.post('/api/v1/credentials/fetch', async (req: express.Request, res: express.Response): Promise<void> => {
        const caller = deps.auth.verify(req.headers['authorization'])
        if (!caller) {
            res.status(401).json({ error: 'unauthorized' })
            return
        }

        const ids = (req.body ?? {}).integration_ids
        if (!Array.isArray(ids) || !ids.every((id) => typeof id === 'number' && Number.isInteger(id))) {
            res.status(400).json({ error: 'integration_ids must be an array of integers' })
            return
        }
        if (ids.length > deps.maxBatchSize) {
            res.status(400).json({ error: `too many integration_ids (max ${deps.maxBatchSize})` })
            return
        }

        const requestId = randomUUID()
        let outcome
        try {
            outcome = await deps.service.getForTeam(caller.teamId, ids)
        } catch (error) {
            logger.error('🔑', '[integration_gateway] credential fetch failed', { requestId, error: String(error) })
            recordFetch(caller.caller, 'error')
            res.status(500).json({ error: 'internal error' })
            return
        }

        // Every requested id appears in the response, resolved or null.
        const integrations: FetchResponseBody['integrations'] = {}
        const resolvedIds: number[] = []
        for (const id of ids) {
            const resolved = outcome.resolved.get(id)
            if (resolved) {
                resolvedIds.push(id)
                integrations[String(id)] = resolved
            } else {
                integrations[String(id)] = null
            }
        }

        emitAudit({
            caller: caller.caller,
            teamId: caller.teamId,
            requested: ids,
            resolved: resolvedIds,
            cacheHits: outcome.cacheHits,
            dbLoaded: outcome.dbLoaded,
            requestId,
        })
        recordFetch(caller.caller, 'ok')

        res.status(200).json({ integrations })
    })

    return router
}
